planner/core/find_best_task.go

```diff
@@ -846,7 +845,105 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 	return task, nil
 }
 
-func (ts *PhysicalTableScan) addPushedDownSelection(copTask *copTask, stats *property.StatsInfo) {
+// pushDownSelAndResolveVirtualCols push some filters down to coprocessor and resolve virtual columns.
+//	1. push down filters and check if there is virtual column.
+//	2. substitute virtual columns in TableScan's Schema to corresponding physical columns.
+//	3. substitute virtual columns in TableScan's filters and push them down.
+//	4. add a Projection upon this DataSource.
+func (ds *DataSource) pushDownSelAndResolveVirtualCols(copTask *copTask, path *accessPath, stats *property.StatsInfo) (t task) {
+	// step 1
+	t = copTask
+	if copTask.indexPlan != nil {
+		ds.addPushedDownIndexScan(copTask, path, stats)
+	}
+	if copTask.tablePlan == nil { // don't need to handle virtual columns in IndexScan
+		return
+	}
+	ts := copTask.tablePlan.(*PhysicalTableScan)
+	if !ds.hasVirtualCol {
+		ds.addPushedDownTableScan(copTask, ts, stats)
+		return
+	}
+
+	// step 2
+	ds.substituteVirtualColumns(ts)
+
+	// step 3
+	for i, expr := range ts.filterCondition {
+		ts.filterCondition[i] = expression.ColumnSubstitute(expr, ds.virtualColSchema, ds.virtualColExprs)
+	}
+	cantBePushed := ts.filterCondition // all Cast cannot be pushed down now
+	ts.filterCondition = ts.filterCondition[:0]
+	ds.addPushedDownTableScan(copTask, ts, stats)
```

Comment:
Since `len(ts.filterCondition)` must is 0 (because last line), this line call `ds.addPushedDownTableScan` is meaningless？

Reply:
updated, PTAL~